src/librustc_codegen_llvm/intrinsic.rs

```diff
@@ -363,6 +365,27 @@ pub fn codegen_intrinsic_call(
                             } else {
                                 bx.lshr(args[0].immediate(), args[1].immediate())
                             },
+                        "rotate_left" | "rotate_right" => {
+                            let is_left = name == "rotate_left";
+                            let val = args[0].immediate();
+                            let raw_shift = args[1].immediate();
+                            if llvm_util::get_version() >= (7, 0) {
+                                // rotate = funnel shift with first two args the same
+                                let llvm_name = &format!("llvm.fsh{}.i{}",
```

Comment:
What are the safety rules on this intrinsic? What happens if the rhs is `9999` or some other nonsense?

Are there any tests that ensure no matter what we throw at the intrinsic we get the same result as the previous naive implementation?

Reply:
These intrinsics are [specified](https://llvm.org/docs/LangRef.html#llvm-fshl-intrinsic) to be well-defined for all inputs. In particular, the shift amount is specified to be modulo the element size, which matches the existing behavior of rotate_left.

There are no direct tests for the intrinsic, but it is tested indirectly through libcore tests (in particular https://github.com/rust-lang/rust/blob/master/src/libcore/tests/num/uint_macros.rs#L65), which also has tests for shift amounts larger than the bit width.

Reply:
Wonderful. Finally a sane intrinsic :laughing: Thanks for the clarification